accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+  void baz() { print('V1.baz'); }
+  void qux() { print('V1.qux'); }
+}
+
+void qux() { print('qux'); }
+
+inline class V2 {
+  final V1 it;
+  V2(this.it);
+  void foo() { print('V2.foo); }
+  void bar() {
+    foo(); // Prints 'V2.foo'.
+    it.foo(); // Prints 'V1.foo'.
+    it.baz(); // Prints 'V1.baz'.
+    1.foo(); // Prints 'E1.foo'.
+    1.baz(); // Compile-time error.
+    qux(); // Prints 'qux'.
+  }
+}
+```
+
+*That is, when the static type of an expression is an inline type `V`
+with representation type `R`, each method invocation on that
+expression will invoke an instance method declared by `V` or inherited
+from a superinterface (or it could be an extension method with on-type
+`V`).  Similarly for other member accesses.*
+
+Let _DV_ be an inline class declaration named `Inline` with type
+parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>s</sub> extends B<sub>s</sub></code>.
+Assume that _DV_ declares a final instance variable with name `id` and
+type `R`.
+
+We say that the _declared representation type_ of `Inline`
+is `R`, and the _instantiated representation type_ corresponding to
+<code>Inline&lt;T<sub>1</sub>,.. T<sub>s</sub>&gt;</code> is
+<code>[T<sub>1</sub>/X<sub>1</sub>, .. T<sub>s</sub>/X<sub>s</sub>]R</code>.
+
+We will omit 'declared' and 'instantiated' from the phrase when it is
+clear from the context whether we are talking about the inline class
+itself, or we're talking a particular instantiation of a generic
+inline. *For non-generic inline classes, the representation type is
+the same in either case.*
+
+Let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+`V` is a proper subtype of `Object?`. If `R` is non-nullable then `V`
+is a proper subtype of `Object` as well.
+
+*That is, an expression of an inline type can be assigned to a top type
+(like all other expressions), and if the representation type is
+non-nullable then it can also be assigned to `Object`. Non-inline types
+(except bottom types) cannot be assigned to inline types without a cast.*
+
+In the body of a member of an inline class declaration _DV_ named
+`Inline` and declaring the type parameters
+<code>X<sub>1</sub>, .. X<sub>s</sub></code>,
+the static type of `this` is
+<code>Inline&lt;X<sub>1</sub> .. X<sub>s</sub>&gt;</code>.
+The static type of the representation name is the representation
+type.
+
+*For example, in `inline class V { final R id; ...}`, `id` has type
+`R`, and `this` has type `V`.*
+
+Again, let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+If `R` is not an inline type then we say that `V` is an inline type
+at level zero. If `R` is an inline type at level _k_ then we say that
+`V` is an inline type at level _k + 1_.
+A compile-time error occurs if the level of `V` is undefined.
+
```

Comment:
Definitely finicky. But perhaps `C` and `Foo` are safe after all?

The erasure step is applied to an inline type, not to an inline class declaration. For any given occurrence of `C<T>` for some `T`, we would be able to check separately that `T` has an inline erasure, and `C<T>` would then have the same inline erasure (because `C` happens to use `T` as the representation type).

In the body of `C`, there is no need to inline erase `T`: At run time, `T` is available in the current scopes, and its value is already inline erased.

If we consider a raw `Foo` then it just means `Foo<C<dynamic>>`, and the inline erasure of that type argument is `dynamic`. If we consider `Foo<S>` for any `S`, then we'd again ("by the induction hypothesis") ensure separately that `S` has an inline erasure.

Reply:
I don't think this flies in general.  It might work out for classes, but even there it makes me nervous.  But with generic methods, these things exist at runtime.  For example, `void Function<T extends C<T>>()`.

Reply:
Did this get handled somewhere that I missed?

Reply:
No, but I did not understand it correctly the first time around. I think this is the single really serious issue that remains in this proposal.

Would it be overkill to require the following?:

Let `X1 extends T1, .. Xs extends Ts` be a type parameter list declaration. A compile-time error occurs if `X1 extends S1, .. Xs extends Ss` has any compile-time errors, where `Sj` is obtained by recursively replacing every occurrence of an inline type in `Tj` by the corresponding instantiated representation type.

In other words, for every type variable declaration, recursively perform an inline-erasure on every inline type in its bound, and then run the normal static analysis checks on it again.

I put this rule into the document.

Reply:
It might be unnecessary to check the inline-erased type parameter bounds of classes/mixins/typedefs/inline_classes, because instance creations are always statically resolved, and hence we can just check the specified bounds at compile time, and trust all type arguments to satisfy the bounds at run time.

However, this won't work for generic functions, because they can be invoked dynamically.

So it's probably not even worth the trouble to make class type parameters an exception.

Another perspective on this is that there are no (real) compile-time errors for type parameter bound declarations: It's an error to have a cycle like `X extends Y, Y extends X`, but we could also just allow it. Similarly for the union type bounds like `X extends FutureOr<X>`: We currently do allow them, but we might make them an error. As long as every actual type parameter satisfies the bounds, it doesn't really matter whether or not there is a cycle among several type parameters.